market_analysis/views/default.py

```diff
@@ -0,0 +1,450 @@
+    return int(shares)
+
+
+def build_stock_entry(
+        y_values, price, shares, value, max_num=None, min_num=None
+        ):
+    """Build a stock entry to be returned to build graph."""
+    return {'y_values': y_values, 'price': price, 'shares': shares,
+            'value': value, 'max': max_num, 'min': min_num}
+
+
+def build_graph(request, elements, percentage=False):
+    """Build and return the graph data from an API request for rendering."""
+    url = 'http://dev.markitondemand.com/MODApis/Api/v2/InteractiveChart/json'
+    req_obj = {     # parameters for API request
+        "parameters":
+        {
+            'Normalized': 'false',
+            'NumberOfDays': 14,
+            'DataPeriod': 'Day',
+            'Elements': elements
+        }
+    }
+    resp = requests.get(url, params=urlencode(req_obj))
+
+    if resp.status_code == 200:
+        total_shares = 0
+        total_value = 0
+        entries = {key: value for key, value in resp.json().items()}    # data from API
+        stocks = {}     # data entries for each stock
+        export = {}     # container of re-packaged data to be rendered
+
+    # build export dict for template
+        export['dates'] = format_dates(entries['Dates'])
+        export['x_values'] = entries['Positions']
+        daily_totals = [0 for j in range(len(export['x_values']))]
+
+        current_user_id = request.dbsession.query(Users).filter(
+            Users.username == request.authenticated_userid
+        ).first().id
+
+        for series in entries['Elements']:
+
+            shares = query_shares(request, current_user_id, series['Symbol'])
+            y_vals = series['DataSeries']['close']['values']
+            price = y_vals[-1]
+
+            if not shares:
+                shares = 0
+
+            total_shares += shares
+            total_value += price * shares
+
+            for i in range(len(y_vals)):
+                daily_totals[i] += (y_vals[i] * shares)
+
+            if percentage:
+                y_vals = prepare_daily_changes(y_vals)
+
+            stocks[series['Symbol']] = build_stock_entry(
+                                    y_vals, price, shares, (price * shares),
+                                    series['DataSeries']['close']['max'],
+                                    series['DataSeries']['close']['min']
+                                    )
+
+        if percentage:
+            daily_change = prepare_daily_changes(daily_totals)
+            stocks['Total'] = build_stock_entry(
+                                    daily_change, round(total_value, 2),
+                                    total_shares, round(total_value, 2)
+                                    )
+
+        export['stocks'] = stocks
+        export['total_shares'] = total_shares
+        export['total_value'] = round((total_value), 2)
+
+        return {'entry': export}
+
+    else:
+        return HTTPFound(request.route_url('api_error'))
```

Comment:
This function isn't a view, so returning a response from it like this feels a bit odd.  It seems like it would be better to allow the function to raise an HTTPError from the requests library, and then catch that exception where the function is used.  It would simplify testing this function by itself.